### PR TITLE
Refactor CoreObject to leverage native JS semantics.

### PR DESCRIPTION
**Pull request by rwjblue**
* Make `.extend()` _actually_ use `class` for proper inheritance
* Move the concepts of `NAME_KEY` to a weakmap
* "prettier" some files...
* prevent meta assertions from eagerly "to-string"ing an object (when the assertion is not going to run anyways)

### CODE REVIEW

**Comment by rwjblue** [comment]
@pzuraq - This is a slightly different approach than your PR, but has basically the same goal. When you are back in the saddle, can you rebase your tests on top of this branch and see how we do?

**Comment by pzuraq** [comment]
👍this looks good based on a quick reading on my phone, will rebase the tests as soon as I’m back!

File: packages/ember-metal/lib/namespace_search.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      processNamespace(namespaces[i]);
+    }
+
+    unprocessedMixins = false;
+  }
+}
+
+export function classToString() {
+  let name = getName(this);
+  if (name !== void 0) {
```
Comment (bekzod) [inline packages/ember-metal/lib/namespace_search.js]: even if it was `undefined` babel would replace it to `void 0`, no ?

Comment (rwjblue) [reply packages/ember-metal/lib/namespace_search.js]: Yes, I believe so

**Comment by rwjblue** [comment]
https://github.com/emberjs/ember.js/pull/16437 enables the testing infrastructure to ensure that we stay compliant...

**Pull request merged**